Broadcast server awareness to all clients

brichet opened the pull request:
~~This PR remove the Awareness to use the one from pycrdt (https://github.com/jupyter-server/pycrdt/pull/170)~~

This PR add broadcasting changes of state in the local awareness.

Draft until https://github.com/jupyter-server/pycrdt/pull/170 is merged and released.

davidbrochart commented on pycrdt_websocket/awareness.py:
```diff
@@ -1,17 +1,65 @@
+from uuid import uuid4
 
-from pycrdt import Decoder, read_message
+import nest_asyncio  # type: ignore[import-untyped]
```
We cannot use `nest-asyncio`, why do you need it?

brichet replied:
Why can't we use it ?

We need it for `loop.run_until_complete(self.set_local_state_field("user", self._user))`, [here](https://github.com/jupyter-server/pycrdt-websocket/pull/73/files#diff-db35ddad107aa5ba618cbd790f840506094190cc99297d1f4aa67a3b7160b395R154).
Otherwise it raises with `event loop already running`.

AFAIU, the `on_change()` callback must be asynchronous to broadcast the message to all users, but the setter must be synchronous.
If we can make [this function](https://github.com/jupyter-server/pycrdt-websocket/pull/73/files#diff-d5f2a35bb64e14b167e41fb7ba849cb6b9cc45197bffb9f11278c457f5848b73R308) synchronous then we can avoid this import.

davidbrochart replied:
`nest-asyncio` monkey-patches `asyncio`, and has other issues.
Anyway, `pycrdt-websocket` is event-loop agnostic as it's based on AnyIO, so we shouldn't use `asyncio` directly. You can probably spawn a task to solve your issue.

brichet replied:
> You can probably spawn a task to solve your issue.

Can you be more specific please ? 

Actually this is what I was trying to do with [this function](https://github.com/jupyter-server/pycrdt-websocket/pull/73/files#diff-d5f2a35bb64e14b167e41fb7ba849cb6b9cc45197bffb9f11278c457f5848b73R308), but I still need to call an async function from a sync one.

brichet replied:
It should be fine actually, thanks.

davidbrochart commented on pycrdt_websocket/awareness.py:
```diff
@@ -56,10 +103,75 @@ def get_changes(self, message: bytes) -> dict[str, Any]:
         }
+
+        # Do not trigger the callbacks if it is only a keep alive update
+        if len(added) or len(filtered_updated) or len(removed):
```
```suggestion
        if added or filtered_updated or removed:
```

davidbrochart commented on pycrdt_websocket/awareness.py:
```diff
@@ -56,10 +103,75 @@ def get_changes(self, message: bytes) -> dict[str, Any]:
+
+    def set_local_state_field(self, field: str, value: Any):
+        current_state = self.get_local_state()
+        current_state.update({field: value})
```
```suggestion
        current_state[field] = value
```

davidbrochart commented on pycrdt_websocket/awareness.py:
```diff
@@ -56,10 +103,75 @@ def get_changes(self, message: bytes) -> dict[str, Any]:
+                return loop.run_until_complete(self.on_change(msg))
+        return None
+
+    def set_local_state_field(self, field: str, value: Any):
```
```suggestion
    def set_local_state_field(self, field: str, value: Any) -> None:
```

davidbrochart commented on pycrdt_websocket/awareness.py:
```diff
@@ -56,10 +97,70 @@ def get_changes(self, message: bytes) -> dict[str, Any]:
+        msg = write_var_uint(YMessageType.AWARENESS) + msg
+
+        if self.on_change:
+            with start_blocking_portal() as portal:
```
But all this happens in the same thread, right? `start_blocking_portal()` is for spawning tasks from a worker thread.

brichet commented:
Thanks @davidbrochart for the review and live debugging.

As suggested in https://github.com/jupyter-server/jupyter_ydoc/pull/277#discussion_r1785810320 I moved the `Awareness` to pycrdt in https://github.com/jupyter-server/pycrdt/pull/170.
So I remove it in this PR, but I can also split it in 2 different PRs, (1) remove `Awareness` and depend on pycrdt's one and (2) make use of `Awareness` in the `YRoom`.

davidbrochart commented on pyproject.toml:
```diff
@@ -30,7 +30,7 @@ classifiers = [
     "anyio >=3.6.2,<5",
     "sqlite-anyio >=0.2.3,<0.3.0",
-    "pycrdt >=0.9.0,<0.10.0",
+    "pycrdt >=0.9.15,<0.10.0",
```
```suggestion
    "pycrdt >=0.9.16,<0.10.0",
```

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -305,7 +305,11 @@ async def serve(self, websocket: Websocket):
             self._handle_exception(exception)
 
-    async def local_update_awareness(self, state):
+    def local_update_awareness(self, state: bytes):
```
```suggestion
    def local_update_awareness(self, state: bytes) -> None:
```

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -305,7 +305,11 @@ async def serve(self, websocket: Websocket):
+        if self._task_group is not None:
+            self._task_group.start_soon(self._local_update_awareness, state)
+
+    async def _local_update_awareness(self, state: bytes):
```
```suggestion
    async def _local_update_awareness(self, state: bytes) -> None:
```

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -305,7 +305,11 @@ async def serve(self, websocket: Websocket):
 
-    async def local_update_awareness(self, state):
+    def local_update_awareness(self, state: bytes):
+        if self._task_group is not None:
```
We should probably raise an exception if it's `None`.

brichet replied:
Done in https://github.com/jupyter-server/pycrdt-websocket/pull/73/commits/44ce8528172bbda2c631960beac7f0b49f834a71

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -304,3 +304,24 @@ async def serve(self, websocket: Websocket):
+            self._task_group.start_soon(self._local_update_awareness, state)
+        else:
+            self.log.error(
+                "Server awareness can't be broadcasted, the task group has not been "
+                "initialized in the YRoom"
```
```suggestion
                "Cannot broadcast server awareness: YRoom not started"
```

brichet commented:
Finally that's a lot of commits and comments for minimal changes :smile:

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -77,11 +78,12 @@ def __init__(
         self.ystore = ystore
         self.log = log or getLogger(__name__)
+        self.awareness = Awareness(self.ydoc)
+        self.awareness.observe(self.local_update_awareness)
```
I'm not sure about the callback name `local_update_awareness`, maybe `send_server_awareness`?

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -304,3 +306,33 @@ async def serve(self, websocket: Websocket):
+            return
+
+        if self._task_group is not None:
+            updated_clients = [
+                *changes[0].get("added", []),
+                *changes[0].get("filtered_updated", []),
+            ]
```
The changes consist of `"added"`, `"updated"` and `"removed"` entries only, not `"filtered_updates"`. Should we include `"removed"`?
```suggestion
            updated_clients = [v for value in changes[0].values() for v in value]
```

brichet replied:
Also I added a test to broadcast only if there is an effective change (not 'update').

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -304,3 +306,30 @@ async def serve(self, websocket: Websocket):
+    def send_server_awareness(self, type: str, changes: tuple[dict[str, Any], Any]) -> None:
+        """
+        Callback to broadcast the server awareness to clients.
+        """
```
```suggestion

        Arguments:
            type: The change type.
            changes: The awareness changes.
        """
```

davidbrochart commented on pycrdt_websocket/yroom.py:
```diff
@@ -304,3 +306,30 @@ async def serve(self, websocket: Websocket):
+        """
+        Callback to broadcast the server awareness to clients.
+        """
+        if type != "change" or changes[1] != "local":
```
Should we actually broadcast every update?
```suggestion
        if type != "update" or changes[1] != "local":
```

brichet replied:
Your right, we should broadcast all events.
Currently the `update` event is triggered only when the `change` event is triggered.
Should we trigger an update every X sec ?

davidbrochart replied:
I think we need [this logic](https://github.com/yjs/y-protocols/blob/255b7e6c96d9c6dd48b5519b3f83ab65afdb4bba/awareness.js#L59) in pycrdt. I'll work on it.

brichet replied:
Yes, that's what I meant.

davidbrochart commented on pyproject.toml:
```diff
@@ -30,7 +30,7 @@ classifiers = [
     "anyio >=3.6.2,<5",
     "sqlite-anyio >=0.2.3,<0.3.0",
-    "pycrdt >=0.9.16,<0.10.0",
+    "pycrdt >=0.10.0,<0.11.0",
```
Let's take the latest bugfix.
I'm also wondering if we should wait for the implementation of periodic awareness updates in pycrdt?
```suggestion
    "pycrdt >=0.10.1,<0.11.0",
```

brichet replied:
I don't think it is required for this change.
But we should probably remove the `type != "update"` condition to broadcast all the event.

brichet replied:
Right, sorry again for the confusion. The `change` event is only for observers, but not for clients...

davidbrochart replied:
OK, I'll merge then.

davidbrochart reviewed (approved):
Thanks @brichet!
